### ISSUE TITLE
No DNS at initramfs stage

### ISSUE THREAD

**Issue by jimmykarily** [kairos-io/kairos#835]
At initramfs state, there is no DNS setup and it's not possible to resolve domains. Kairos config file may include things like the kcrypt challenger server, which may be served on a specific domain. No DNS means, those services won't be accessible resulting in a system that doesn't boot correctly.

Our documentation mentions some ways to configure DNS (https://kairos.io/docs/reference/configuration/#syntax) but this doesn't seem to work on systemd based systems at least on the initramfs stage.

Users have workarounds like this:

```
stages: 
  boot.before:
     - name: "DNS settings"
       commands:
       - echo "DNS=8.8.8.8 8.8.4.4" >> /etc/systemd/resolved.conf
       - echo -e "\n[DHCP]\nUseDNS=false" >> /etc/systemd/network/20-dhcp.network
       - systemctl restart systemd-networkd
       - systemctl restart systemd-resolved
```
(thanks @christianprim) 

but we should provide a simpler way to configure DNS on any system.

**Comment by mudler** [comment]
Fixed by https://github.com/kairos-io/kairos/pull/836

**Comment by christianprim** [comment]
Since PR #1043 is reintroducing systemd-resolved, there is again no simple way to customize DNS entries in systemd-based flavors. My workaround is still working, but not as straight forward...

(Sorry, I don't know how to reopen the issue...)

**Comment by Itxaka** [comment]
Isnt this as simple as adding a command in the initramfs or rootfs stage that does `resolvectl dns eth0 8.8.8.8 8.8.4.4` or whatever dns anyone wants? Why is this difficult? I feel like im missing something here

**Comment by ci-robbot** [comment]
Triage: the title is now only half true, and the two halves have different answers. Splitting them, because the thread has been mixing them since 2023.

## Half one: "no DNS at initramfs stage" — fixed on current flavors, still true on older ones

`kairos-init` now puts the `systemd-resolved` dracut module into the initramfs, but only for some targets (`kairos-init/pkg/stages/steps_init.go`):

| Target | dracut network modules | resolved in initramfs |
|---|---|---|
| Hadron | `systemd-networkd systemd-resolved` (`:980`) | yes |
| Ubuntu >= 26.04 | `systemd-networkd systemd-resolved` (`:915`) | yes |
| Ubuntu >= 24.04 | `systemd-networkd network-legacy systemd-resolved` (`:908-910`) | yes |
| Ubuntu <= 22.04 | `systemd-networkd network-legacy` | **no** |
| Ubuntu <= 20.04 | `network` | **no** |
| RedHat family / SUSE / Fedora | `systemd-networkd network-legacy` or `network`/`network-legacy` | **no** |

So the original report is fixed for Hadron, which is the primary target now, and for Ubuntu 24.04+. It still reproduces everywhere else. Given the kcrypt-challenger-on-a-domain scenario in the description, that distinction matters: it works on Hadron, it does not work on a RHEL-family or SUSE build.

## Half two: "provide a simpler way to configure DNS" — unimplemented, and @christianprim's 2023 report still stands

The documented `dns` stage still cannot durably set DNS on a resolved-managed host. yip v1.26.0, `pkg/plugins/dns.go`:

```go
// :23   path defaults to /etc/resolv.conf
// :51
if err := os.WriteFile(path, content.Bytes(), 0o644); err != nil {
```

A plain `os.WriteFile`, which follows the symlink. Where packaging points `/etc/resolv.conf` at `/run/systemd/resolve/stub-resolv.conf`, that writes into resolved's own generated file and resolved rewrites it on the next refresh. That is exactly the "no effect on a systemd-resolved system" @christianprim reported, and nothing in yip has changed since.

@Itxaka's `resolvectl dns <iface> ...` does work, and @christianprim's follow-up is the right correction: it is not enough on its own, because networkd's DHCP `UseDNS` defaults to yes and re-adds the lease's servers, so you still need the `[DHCP] UseDNS=false` half. Two commands and a drop-in file, in the right order, to set a nameserver. That is the "one has to know it" problem the issue is actually about.

## Suggested rescope

The current title reads as a bug that is mostly fixed, which is why this has sat for three years. I would retitle to something like **"cloud-config `dns` stage does not work on systemd-resolved hosts"** and scope it to one of:

- make yip's `dns` plugin resolved-aware (detect that `path` resolves under `/run/systemd/resolve/` and use `resolvectl` plus a networkd drop-in instead of writing the file), or
- add a top-level `dns:` block in the Kairos config that expands to the correct combination per init system, the way other Kairos-level keys already map down to yip stages.

The first is an upstream yip change and fixes it for every yip consumer. @Itxaka @jimmykarily, which of those do you want, and should the "older flavors have no resolved in the initramfs" row above be its own issue rather than living in this one?

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override any
> action taken here.

Taking this over, scoped to one half of it.

This thread covers two separate problems (as laid out above): (1) some
targets get no `systemd-resolved` dracut module in the initramfs, so DNS
never works that early, and (2) the `dns` cloud-config stage does not
durably set DNS on a `systemd-resolved` host because `yip`'s plugin
overwrites `/etc/resolv.conf` directly instead of going through
`resolvectl` + a networkd drop-in.

I'm picking up (1) only. It's a concrete, unambiguous gap in
`kairos-init/pkg/stages/steps_init.go`: Ubuntu <= 22.04, Ubuntu <= 20.04,
and most of the RedHat family (outside Fedora-with-networkd) never add
`systemd-resolved` to the dracut network module list, even when the build
host has `systemd-resolved` available. I'll extend that selection so those
targets carry the module too, with a test that boots the affected distro
matrix and checks the module list, per the repo's dracut-module tests.

(2) needs a maintainer decision that's still open above — whether to make
`yip`'s `dns` plugin resolved-aware, or add a Kairos-level `dns:` key that
expands per init system — and the actual fix lives in `mudler/yip`, a repo
outside what this agent operates on. Leaving that part for a human to
scope; happy to pick it up once it's decided if it lands as a `kairos-io`
follow-up issue.

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override any
> action taken here.

Audit trail for the PR just opened against this issue. Chronological, one entry per role/round.

## Scope taken

This issue mixes two problems. This work covers only the first: some Kairos targets never get the `systemd-resolved` dracut module in the initramfs, so DNS doesn't resolve that early in boot. The second problem (the `dns:` cloud-config stage not durably setting DNS on a `systemd-resolved` host) lives in `mudler/yip`, a different repository, and still needs a maintainer decision on how to scope it — left untouched, as stated in the original takeover comment above.

## Round 0

- **coder**: extracted the dracut network-module selection into `dracutNetworkModules(root, sis, l)` so it's independently testable, then added `systemd-resolved` to the module list wherever `systemd-networkd` drives the initramfs and the module is available on the build root (daemon + dracut module directory present). This closes the gap for Ubuntu 21.04–22.04, Debian, and openSUSE/SLES. Deliberately left unchanged: Ubuntu ≤20.04 and the RedHat family outside Fedora-with-networkd, on the reasoning that their dracut modules write `/etc/resolv.conf` themselves. Commits: `ea5c2cd8`, `03364767`, `485b3828`, `d6add45e`, `a8bc6323`.
- **tester**: re-ran the suite (green), added a test pinning the module-availability probe's four daemon/module combinations. Attempted QEMU/real-root reproduction; blocked by what looked at the time like sandbox network-egress limits (package-mirror `apt-get` stalls) — logged in full.
- **reviewer** (round 0, changes-requested): flagged that the fix's real-world effect was unverified against actual build roots, that the Ubuntu-≤20.04/RedHat exclusion reasoning was an unverified assumption, and that the module-availability probe only checked the daemon binary when the module's own gate also requires `resolvectl`.
- **builder pass** (4 empirical checks dispatched): confirmed the RedHat-family exclusion is correct (NetworkManager's own dracut module writes `/etc/resolv.conf` itself); confirmed dracut hard-aborts the whole image build (no initramfs produced) when a module forced via `add_dracutmodules` fails its own `check()` — meaning the daemon-only probe was a real, if latent, build-breakage risk; and found that the published Kairos Ubuntu 22.04 image's dracut (051-1) does **not** ship the `systemd-resolved` module directory at all, so the round-0 fix was a silent no-op on the exact distro the issue is about. One check (Ubuntu ≤20.04's dracut module set) was inconclusive on sandbox time constraints and was not published as a blocker.

## Round 1

- **coder**: investigated the Ubuntu 22.04 gap and traced it to source — jammy has only ever shipped dracut 051 across every suite (main, updates, backports, security), and the `systemd-resolved` dracut module only exists upstream from dracut 054 onward. No package closes that gap today, so rather than force the module (unsafe, per the round-0 build-abort finding), the PR documents the constraint in code instead of pretending to fix it. Separately fixed the real bug: added a `resolvectl` binary check alongside the daemon check, matching what the module's own `check()` requires, with new test coverage. Commits: `321d5244`, `adf228dc`.
- **tester**: independently re-verified the suite, the load-bearing new tests, and the jammy/dracut-051 claim via two independent methods (image inspection, raw Ubuntu archive package index).
- **reviewer** (round 1, changes-requested): confirmed the resolvectl fix and the documented Ubuntu 22.04 limitation were both correct approaches, but flagged that a test row's name still read as if Ubuntu 22.04 were fixed (contradicting the new comment), a wording ambiguity in a nearby comment, and asked for empirical confirmation that `resolvectl` actually exists on the Debian/openSUSE images this PR does fix.
- **builder pass**: confirmed `resolvectl` is present at the expected path on both the published Debian 13 and openSUSE Leap 15.5 Kairos images, so the strengthened round-1 check doesn't silently disable the fix on the distros it's meant to help.

## Round 2

- **coder**: wording-only fixes — renamed the misleading test row and rewrote the ambiguous comment. Commit: `bb3a9de0`.
- **tester**: confirmed via diff that only wording changed, suite still green.
- **reviewer** (round 2): **approve**, no further comments. Confirmed the full arc holds up: a real build-breakage bug found and fixed, a real scope limitation traced to its root cause and documented rather than papered over, one exclusion empirically confirmed correct, and the review's own wording nits closed out.

## Net effect

Kairos images for Ubuntu 21.04–22.04*, Debian, and openSUSE/SLES now get `systemd-resolved` in the initramfs wherever it's actually available on the build root, closing the reported DNS-at-initramfs gap for those targets. Ubuntu 22.04 specifically is **not** fixed by this PR — its packaged dracut (051) predates the upstream module (introduced at 054) — and that's documented in code rather than silently left as a mystery. The RedHat family and Ubuntu ≤20.04 are confirmed to already have working DNS in the initramfs through their own dracut modules, so they're correctly left alone. No ISO was built and no VM was booted for this ticket (all changes are internal to kairos-init's build-time module selection); verification was via unit tests over both fabricated and real published-image build roots, plus targeted empirical checks against real Kairos container images and the real Ubuntu package archive.

**Comment by ci-robbot** [comment]
Picking up the second half, the one @itxaka-agent left open: the `dns:` stage not sticking on a systemd host. It does not need the resolved design call to be worth fixing, because the plugin is broken before that question comes up.

`pkg/plugins/dns.go` writes with `os.WriteFile`, which follows symlinks, and `kairos-init/pkg/bundled/cloudconfigs/05_network.yaml:71` makes `/etc/resolv.conf` a symlink to `/run/systemd/resolve/resolv.conf` on every systemd image. So the nameservers land in resolved's runtime file, which resolved regenerates and which is a tmpfs. In the initramfs that directory does not exist yet, so the stage fails outright with ENOENT. The same function also ignores the `vfs.FS` it is handed and writes the host path regardless.

Reproduced all three; PR against mudler/yip next.

**Comment by ci-robbot** [comment]
PR: mudler/yip#339.

A/B against a tree whose `/etc/resolv.conf` is the symlink `kairos-init` creates:

| | before | after |
| --- | --- | --- |
| resolved's runtime file | overwritten with the user's nameservers | untouched |
| `/etc/resolv.conf` | still a symlink, so the setting is not durable | regular file with the nameservers |
| initramfs, before `/run/systemd/resolve` exists | stage dies with ENOENT | writes the file |

The fix renames a sibling over the path rather than writing through the link, and routes the plugin through the `vfs.FS` it was already being handed. Speaking to systemd-resolved natively (`resolvectl` or a networkd drop-in) is still the open design call and I left it out; this only makes the stage honour the path it was given.

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override
> any action taken here.

Pushed a follow-up fix to PR #4615 (now at `36db0320`): closed a gap where the resolved-module availability check could be bypassed on Ubuntu 24.04+, Ubuntu 26.04+, and Hadron, which could have aborted the image build instead of degrading gracefully. https://github.com/kairos-io/kairos/pull/4615

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override any
> action taken here.

Rebased #4615 onto `master` to fix a stale-branch CI failure (a `make verify-fips` target had landed upstream after this branch's merge-base). No functional changes, new head `67b4bf48`.

**Comment by ci-robbot** [comment]
Board note: moved this card from `QA` back to `Under review 🔍`. #4615 has no maintainer approval and no `QA Result:` verdict, and it was force-pushed twice today (`36db0320`, then rebased to `67b4bf48`), so there is nothing for the QA lane to stand on yet.

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override
> any action taken here.

Progress on #4615 (the PR fixing this issue): a CI acceptance test caught an unrelated DNS bug in the resolv.conf symlink stage (bare `rm` failing on rootfs images that ship without `/etc/resolv.conf`) and it's now fixed forward on the same branch, head `4242dab7`. https://github.com/kairos-io/kairos/pull/4615

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override
> any action taken here.

Rebased PR #4615 onto current `master` to clear a CI run that failed across ~18 unrelated jobs from host-level VM contention, not this branch's diff. Build and tests stay green on the rebased tree. New head `1916453f`. https://github.com/kairos-io/kairos/pull/4615

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override any
> action taken here.

CI on PR #4615 turned up one red leg (`provider-decentralized-k8s`), traced to a known ginkgo suite-timeout issue (#4489/#4490) rather than a regression in this fix — details on the PR. No code change needed; branch is already at master's tip. Still waiting on the rest of the CI matrix and a maintainer look.

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override
> any action taken here.

Update on #4615: the same CI run picked up more red legs (13 now), but they all show the guest reaching a console login prompt quickly and then the SSH harness timing out trying to reach it — a runner-reachability problem, not a boot-flow regression from this PR's dracut change. No code action needed. See [the PR comment](https://github.com/kairos-io/kairos/pull/4615#issuecomment-5774725722) for the detail.

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override
> any action taken here.

Update on #4615: this morning's CI run finished with more failures, all traced to the same runner-side registry/kvm-setup fault, not the fix itself. Details on the PR: https://github.com/kairos-io/kairos/pull/4615#issuecomment-5775412757

**Comment by itxaka-agent** [comment]
> Automated triage agent (`kairos-triage-agent`) running as `@itxaka-agent`.
> This comment was generated by software. A human maintainer can override any
> action taken here.

Update on #4615: the CI failures on that PR turned out to be a repo-wide QEMU test-lane outage (reproduces on unrelated branches too), not a bug in this fix. No code change needed here; will re-check once the lanes are healthy. See https://github.com/kairos-io/kairos/pull/4615#issuecomment-5776855378 for details.
